### PR TITLE
iterable interface for NSArray (and subclasses).

### PR DESCRIPTION
**Pull request by Longhanks**
As requested in #15 

Implements a pythonic interface for NSArray and its subclasses.

Examples from the unit test:

```python
NSMutableArray = ObjCClass('NSMutableArray')
ns_array = NSMutableArray.array()
test_list = ['item1', 'item2', 'item3', 'item4', 'item5']
for item in test_list:
    ns_array.addObject_(item)

# Todo: implement len
self.assertEqual(ns_array.count, 5)

# NSArray itself can be compared to other NSArrays
# NSArray can be copied via copy-slice [:]
self.assertEqual(ns_array[:], ns_array)

# Needed because an NSArray does NOT equal a Python list. Their elements can, however, be equal.
def compare(some_NSArray, python_list):
    self.assertEqual([item for item in some_NSArray], python_list)

# Object retrieval via index
self.assertEqual(ns_array[3], test_list[3])

# positive indices
compare(ns_array[1:], test_list[1:])
compare(ns_array[2:3], test_list[2:3])
compare(ns_array[4:2], test_list[4:2])
compare(ns_array[:4], test_list[:4])

# negative indices
compare(ns_array[-1:], test_list[-1:])
compare(ns_array[-2:-3], test_list[-2:-3])
compare(ns_array[-4:-2], test_list[-4:-2])
compare(ns_array[:-4], test_list[:-4])

# positive indices with step 2
compare(ns_array[1::2], test_list[1::2])
compare(ns_array[2:3:2], test_list[2:3:2])
compare(ns_array[4:2:2], test_list[4:2:2])
compare(ns_array[:4:2], test_list[:4:2])

# negative indices with step 2
compare(ns_array[-1::2], test_list[-1::2])
compare(ns_array[-2:-3:2], test_list[-2:-3:2])
compare(ns_array[-4:-2:2], test_list[-4:-2:2])
compare(ns_array[:-4:2], test_list[:-4:2])

# positive indices with step -3
compare(ns_array[1::-3], test_list[1::-3])
compare(ns_array[2:3:-3], test_list[2:3:-3])
compare(ns_array[4:2:-3], test_list[4:2:-3])
compare(ns_array[:4:-3], test_list[:4:-3])

# negative indices with step -3
compare(ns_array[-1::-3], test_list[-1::-3])
compare(ns_array[-2:-3:-3], test_list[-2:-3:-3])
compare(ns_array[-4:-2:-3], test_list[-4:-2:-3])
compare(ns_array[:-4:-3], test_list[:-4:-3])

# test string representation
self.assertEqual(repr(ns_array), 'NSArray ' + repr(test_list))
```

The one thing I can't get to work is `__len__`, implementing it confuses `core_foundation.py to_value` and the Objective C classes can't be fetched from the runtime anymore... any hint would be great!

### CODE REVIEW

File: rubicon/objc/objc.py [hunk 1]
```diff
@@ -20,6 +20,7 @@
 ######################################################################
 
 objc = cdll.LoadLibrary(util.find_library('objc'))
+Foundation = cdll.LoadLibrary(util.find_library('Foundation'))
```
Comment (freakboy3742) [inline rubicon/objc/objc.py]: Leaving out Foundation was a strategic choice; I was trying to avoid the core of Rubicon spiralling into a a huge project. If it's necessary, then we will include it, but I'd like to explore if we can do this with CFArray (from CoreFoundation)

File: rubicon/objc/objc.py [hunk 2]
```diff
@@ -1287,7 +1288,7 @@ def __setattr__(self, name, value):
         if method:
             ObjCBoundMethod(method, self.__dict__['ptr'])(value)
         else:
-            self.__dict__[name] = value
+            super(ObjCClass, self).__setattr__(name, value)
```
Comment (freakboy3742) [inline rubicon/objc/objc.py]: This seems like a red flag to me; what's the purpose behind this change?

File: rubicon/objc/objc.py [hunk 3]
```diff
@@ -1324,6 +1325,10 @@ def __new__(cls, object_ptr):
         class_ptr = c_void_p(objc.object_getClass(object_ptr))
         objc_instance.__dict__['objc_class'] = ObjCClass(class_ptr)
 
+        # If it's an array, we need the internal counter in the instance to implement iterable.
```
Comment (freakboy3742) [inline rubicon/objc/objc.py]: No you don't - because that makes the iterable object stateful. A ListIterator is a separate object from a list, so you can perform 2 parallel iterations over the same list.

File: tests/test_rubicon.py [hunk 4]
```diff
@@ -428,3 +428,66 @@ def test_function_NSEdgeInsetsMake(self):
         self.assertEqual(insets.bottom, other_insets.bottom)
         self.assertEqual(insets.right, other_insets.right)
 
+    def test_NSArray_iterable(self):
+        "Pythonic interface for NSArray (and its subclasses) as iterable."
```
Comment (freakboy3742) [inline tests/test_rubicon.py]: This looks like it's just a test for `NSMutableArray`, not `NSArray`.

File: tests/test_rubicon.py [hunk 5]
```diff
@@ -428,3 +428,66 @@ def test_function_NSEdgeInsetsMake(self):
...
+
+        NSMutableArray = ObjCClass('NSMutableArray')
+
+        ns_array = NSMutableArray.array()
+        test_list = ['item1', 'item2', 'item3', 'item4', 'item5']
+        for item in test_list:
+            ns_array.addObject_(item)
+
+        # Todo: implement len
```
Comment (freakboy3742) [inline tests/test_rubicon.py]: Intrigued to know what problem you encountered here.

File: tests/test_rubicon.py [hunk 6]
```diff
@@ -428,3 +428,66 @@ def test_function_NSEdgeInsetsMake(self):
...
+        compare(ns_array[:4:-3], test_list[:4:-3])
+
+        # negative indices with step -3
+        compare(ns_array[-1::-3], test_list[-1::-3])
+        compare(ns_array[-2:-3:-3], test_list[-2:-3:-3])
+        compare(ns_array[-4:-2:-3], test_list[-4:-2:-3])
+        compare(ns_array[:-4:-3], test_list[:-4:-3])
+
+        # test string representation
+        self.assertEqual(repr(ns_array), 'NSArray ' + repr(test_list))
```
Comment (freakboy3742) [inline tests/test_rubicon.py]: This is a fine test of the capabilities of the Python interface when the object has been created Python-side; but it doesn't check if an object that has been received through an API will acquire those methods. To test this, you'll need to create a method on the test harness that returns `NSArray`/`NSMutableArray`.

File: rubicon/objc/objc.py [hunk 7]
```diff
@@ -1324,6 +1325,10 @@ def __new__(cls, object_ptr):
...
+        if objc.class_getName(class_ptr).decode('utf-8').startswith('__NSArray'):
```
Comment (dimaqq) [inline rubicon/objc/objc.py]: What about `NSMutableArray`? Does it not belong to same category? And how about `NSDictionary`

File: rubicon/objc/objc.py [hunk 8]
```diff
@@ -1342,14 +1347,69 @@ def __new__(cls, object_ptr):
...
+                    key = self.count - (-1 * key)
+                if key > (self.count - 1) or key < 0:
+                    raise IndexError('list index out of range')
+                else:
+                    return self.objectAtIndex_(key)
+            elif isinstance(key, slice):
+                array = ObjCClass('NSMutableArray').array()
+                helper = [item for item in self]
+                for item in helper[key]:
+                    array.addObject_(item)
```
Comment (dimaqq) [inline rubicon/objc/objc.py]: This design decision needs discussion:

If `isinstance(a, NSArray)`, would you rather `a[:]` return a Python list or clone `NSArray`?

**Comment by freakboy3742** [comment]
This has been superseded by #42.
